src/vos/vos_dtx.c

```diff
@@ -1229,15 +1230,16 @@ vos_dtx_check_availability(daos_handle_t coh, uint32_t entry,
 		return ALB_UNAVAILABLE;
 	}
 
-	/* For rebuild fetch, if hit non-committed DTX, has to ask the
-	 * sponsor to wait and retry, otherwise, the new in-rebuilding
-	 * target may miss some data.
-	 * Under such case, DTX resync on related leader server is not
-	 * finished yet. The DTX entry may be stale, need refresh with
-	 * the leader.
+	/*
+	 * Up layer rebuild logic guarantees that the rebuild scan will not be
+	 * triggered until DTX resync has been done on all related targets. So
+	 * here, if rebuild logic hits non-committed DTX entry, it must be for
+	 * new IO that version is not older than rebuild, then it is invisible
+	 * to rebuild. Related new IO corresponding to such non-committed DTX
+	 * has already been sent to the in-rebuilding target.
 	 */
 	if (intent == DAOS_INTENT_MIGRATION)
-		return dtx_inprogress(dae, dth, false, true, 6);
+		return ALB_UNAVAILABLE;
```

Comment:
maybe worth a warning if the version is older than rebuild, in case this will cover some other bugs.

Reply:
The original logic is for old rebuild, at that time, each DAOS target did each own DTX resync and rebuild independently. That is why we need to check the visibility in VOS for rebuild. But after you introduced global DTX resync version, the rebuild will wait DTX resync tp complete on all targets, then when VOS visibility check comes here, every rebuild visible DTX must have been handled by the completed DTX resync, so the left ones are invisible to rebuild.